main.cc

```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
+#include <cassert>
+
+using namespace std;
+
+class String {
+public:
+    
+    String() :
+        String("", 0)
+    { }
+
+    String(const char* data) :
+        String(data, strlen(data))
+    { }
+
+    String(const String& s) :
+        String(s.data(), s.length())
+    { }
+    
+    String(const char* data, size_t length) {
+        alloc_and_copy(data, length);
+    }
+    
+    String(unsigned capacity) {
```

Comment:
Although it was not stated in the assignment explicitly, we'd like to mimic the standard container behavior. If a standard container's constructor get an integer, it is the desired *size* (not the capacity).   In order to set capacity, containers define `reserve` member function.

Reply:
But in this case, if it is the size and not the capacity, I am left with un-initialized data?

Reply:
You'll have _value-initialized_ data (please, read about value-initialization).